src/vm/analysis/trait_checker/post_type_check.rs

```diff
@@ -0,0 +1,80 @@
+        command.run(contract_analysis, analysis_db)?;
+        Ok(())
+    }
+}
+
+impl PostTypeCheckingTraitChecker {
+
+    fn new() -> Self {
+        Self {
+        }
+    }
+
+    pub fn run(&mut self, contract_analysis: &mut ContractAnalysis, analysis_db: &mut AnalysisDatabase) -> CheckResult<()> {
+    
+        for trait_identifier in &contract_analysis.implemented_traits {
+
+            let trait_name = trait_identifier.name.to_string();
+            let contract_defining_trait = analysis_db.load_contract(&trait_identifier.contract_identifier)
+                .ok_or(CheckErrors::TraitReferenceUnknown(trait_identifier.name.to_string()))?;
+            let trait_sig = contract_defining_trait.get_defined_trait(&trait_name)
+                .ok_or(CheckErrors::TraitReferenceUnknown(trait_identifier.name.to_string()))?;
+
+            for (func_name, expected_sig) in trait_sig.iter() {
+                match contract_analysis.get_public_function_type(func_name) {
+                    Some(FunctionType::Fixed(func)) => {
+                        if func.args.len() != expected_sig.args.len() {
+                            return Err(CheckErrors::BadTraitImplementation(trait_name.clone(), func_name.to_string()).into())
+                        }
+                        let args = expected_sig.args.iter().zip(func.args.iter());
+                        for (expected_arg, arg) in args {
+                            match (expected_arg, &arg.signature) {
+                                (TypeSignature::TraitReferenceType(expected), TypeSignature::TraitReferenceType(actual)) => {
+                                    let expected_trait_id = contract_defining_trait.get_referenced_trait(&expected.to_string())
+                                        .ok_or(CheckErrors::BadTraitImplementation(trait_name.clone(), func_name.to_string()))?;
+                                    let actual_trait_id = contract_analysis.get_referenced_trait(&actual.to_string())
+                                        .ok_or(CheckErrors::BadTraitImplementation(trait_name.clone(), func_name.to_string()))?;
+                                    if actual_trait_id != expected_trait_id {
+                                        return Err(CheckErrors::BadTraitImplementation(trait_name.clone(), func_name.to_string()).into())
+                                    }
+                                }
+                                _ => {
+                                    if !expected_arg.admits_type(&arg.signature) {
+                                        return Err(CheckErrors::BadTraitImplementation(trait_name.clone(), func_name.to_string()).into())
+                                    }        
+                                }
+                            }
+                        }
+
+                        if !expected_sig.returns.admits_type(&func.returns) {
+                            // todo(ludo): investigate (response uint uint) admiting (response uint)
```

Comment:
Not sure what this comment means.  `(ok u1)` and `(err u2)` should be admitted by `(response uint uint)`, for example.

Reply:
Leftover - striped with https://github.com/blockstack/blockstack-core/pull/1246/commits/b676529c95c620c267137d026fff467336fbbabd